src/pages/workspaces/workspace/useWorkspace.ts

```diff
@@ -0,0 +1,186 @@
+import _ from 'lodash/fp'
+import { Fragment, useRef, useState } from 'react'
+import { div, h } from 'react-hyperscript-helpers'
+import { Link } from 'src/components/common'
+import { locationTypes } from 'src/components/region-common'
+import { updateRecentlyViewedWorkspaces } from 'src/components/workspace-utils'
+import { Ajax } from 'src/libs/ajax'
+import { responseContainsRequesterPaysError } from 'src/libs/ajax/ajax-common'
+import { saToken } from 'src/libs/ajax/GoogleStorage'
+import { getConfig } from 'src/libs/config'
+import { withErrorIgnoring, withErrorReporting } from 'src/libs/error'
+import { clearNotification, notify } from 'src/libs/notifications'
+import { useCancellation, useOnMount, useStore } from 'src/libs/react-utils'
+import { getUser, workspaceStore } from 'src/libs/state'
+import * as Utils from 'src/libs/utils'
+import { differenceFromNowInSeconds } from 'src/libs/utils'
+import { isAzureWorkspace, isGoogleWorkspace, WorkspaceWrapper } from 'src/libs/workspace-utils'
+import { defaultLocation } from 'src/pages/workspaces/workspace/analysis/runtime-utils'
+
+
+interface StorageDetails {
+  googleBucketLocation: string // historically returns defaultLocation if cannot be retrieved or Azure
+  googleBucketType: string // historically returns locationTypes.default if cannot be retrieved or Azure
+  azureContainerRegion: string|undefined
+  azureContainerUrl: string|undefined
+  azureContainerSasUrl: string|undefined
+}
+
+export type InitializedWorkspaceWrapper = WorkspaceWrapper & { workspaceInitialized: boolean }
+
+interface WorkspaceDetails {
+  workspace: InitializedWorkspaceWrapper
+  accessError: boolean
+  loadingWorkspace: boolean
+  storageDetails: StorageDetails
+  refreshWorkspace: () => {}
+}
+
+export const googlePermissionsRecheckRate = 15000
+export const azureBucketRecheckRate = 5000
+
+export const useWorkspace = (namespace, name) : WorkspaceDetails => {
```

Comment:
Can you explain a bit this refactor into `useWorkspace` is doing? Is this purely for testability and making things more manageable or are there other reasons?

Reply:
`WorkspaceContainer` is very complicated-- it does a lot of wrapping of things (adding the menu, context bar, storing state, making Ajax calls, etc.). By pulling the state management/Ajax calls mostly into a hook function, it helps reduce the complexity of `WorkspaceContainer`, and it makes the hook function code testable. Ideally the hook function would have no HTML elements at all, but there is a bit of legacy code in there that is showing a notification (though TBH I really suspect that code is dead and could just be deleted…. but I'm trying to be somewhat conservative with the refactoring).

We had no unit test coverage at all for `WorkspaceContainer`, and this allows much of the more complicated code to be tested. It also is one step towards converting it all to typescript.